### ISSUE TITLE
apply only subsets of the built resources.

### ISSUE THREAD

**Issue by benjamin-bergia** [kubernetes-sigs/kustomize#1337]
I am currently having the following workflow with structure and workflow with my kustomize project. I have a base containing all my apps and overlays for each environments (prod, staging, etc..). Each app contains a base with usually a deployment and some base settings and overlays for different variations of the app.

In my CD pipeline I run `kustomize build  myEnv | kubectl apply -f -` and everything is fine.
From time to time, we make breaking changes that require to apply changes step by step. Like for example creating a new statefulset, copying some data before touching the service. This cannot just be applied automatically. 

Unfortunately, since I have basically two layers of overlay (app variant and environment), I cannot build a single resource at the time. I would have to manually remove all resources from my base kustomization.yml in order to manually control what is built. My work around at the moment is to pipe `kustomize build` to a script that split each resources in it's own file based on the combination of namespace/name/kind. I can then apply the specific files in the order that are needed. An added benefit is that this files can be version controlled and change results are obvious during PRs.

This might be an anti-pattern, I am not sure, but I would really to be able to build and then be able to easily choose what to apply. Is there any thought on this?

**Comment by jbrette** [comment]
@benjamin-bergia You can kind of achieve the filtering by using the -o <outputdir>.  This only requires to create the directory ahead of time otherwise is will create the <outputdir> file.

```bash
mkdir outputdir
kustomize build myEnv -o outputdir
kubectl apply -f outputdir/<xxxx>.yaml
```

**Comment by benjamin-bergia** [comment]
Hi @jbrette,
Yes it does create an output file but this file will contain all the resources generated. You then have no easy way to apply only some of these resources.

**Comment by looztra** [comment]
Shameless plug to [kubesplit](https://github.com/looztra/kubesplit) that I coded because I had the same need : 

```bash
kustomize build myEnv | kubesplit -o outputDir -q
```
`-o` to specify the output directory, `-p` to not use a file file resource prefix [or don't use it and you'll get ordered file list]

This will generate one file per resource, clusterwide resources  at the root, namespaced resources in a directory matching the namespace name.

**Comment by benjamin-bergia** [comment]
@looztra Perfect!

**Comment by farthinder** [comment]
I didnt have this exact use case, but similar and found a need solution. 

I needed to deploy a subset of resources during the initial deployment and then deploy everything once I had longhorn in place. I added a label to the resources that needed to be applied first and solved it with yq:

`kustomize build . | yq eval -P | envsubst | yq e 'select(.metadata.labels.initialDeployment == "true")' - | kubectl apply -f -`


`yq eval -P | envsubst` is likely not needed in all use cases.
